pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpumanager
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/state"
+	"k8s.io/kubernetes/pkg/kubelet/status"
+)
+
+type kletGetter interface {
+	GetPods() []*v1.Pod
+	GetCachedMachineInfo() (*cadvisorapi.MachineInfo, error)
+}
+
+type runtimeService interface {
+	UpdateContainerResources(id string, resources *runtimeapi.LinuxContainerResources) error
+}
+
+type policyName string
+
+// Manager interface provides methods for kubelet to manage pod cpus
+type Manager interface {
+	Start()
+
+	// RegisterContainer is called between container creation and container start
+	// so that initial CPU affinity settings can be written through to the
+	// container runtime before the first process begins to execute.
+	RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error
+
+	// UnregisterContainer is called near UnregisterContainer() so that the manager
+	// stops trying to update that container in the reconcilation loop and
+	// any CPUs dedicated to the container are freed to the shared pool.
+	UnregisterContainer(containerID string) error
```

Comment:
Similarly, I think this should be called PreRemoveContainer since it is called after logs are removed, and before the container is itself removed.  If we align the language of the manager around those terms, we could pass in a ContainerLifecycleHandler style interface into the runtime syncPod.

I think the current naming schemes confuse what the manager is (a manager of state) versus what the caller perceives it to be (a lifecycle hook that should be called during life of container).

WDYT?

Reply:
Makes sense to me. Would slightly prefer something like `PostStopContainer` since that's the more significant order we care about (pre-remove could be called almost anytime, in comparison.)

Reply:
Will be happy to make these naming changes, but they will appear in PR #51140 first (know it's difficult to keep track of dependent patch sets on GH)